[JSC] Use `enum class` for `CompilationResult`, `CachedCodeBlockTag`, `CodeSpecializationKind` for Safer C++

Gumichocopengin8 opened the pull request:
#### a2ae5819255e50ce1ec7347fa8b708292d61ecf6
<pre>
[JSC] Use `enum class` for `CompilationResult`, `CachedCodeBlockTag`, `CodeSpecializationKind` for Safer C++
<a href="https://bugs.webkit.org/show_bug.cgi?id=295459">https://bugs.webkit.org/show_bug.cgi?id=295459</a>

Reviewed by Yijia Huang and Keith Miller.

This is a refactoring-only patch with no behavior changes.
Use `enum class` for `CompilationResult`, `CachedCodeBlockTag`, `CodeSpecializationKind` for safer C++ [1].

[1]: <a href="https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines#use-enum-classes-instead-of-old-style-enumerations">https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines#use-enum-classes-instead-of-old-style-enumerations</a>

Canonical link: <a href="https://commits.webkit.org/297204@main">https://commits.webkit.org/297204@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/26b52eadbb5860bededc559ff281211b6357f351

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110863 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30526 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20960 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116893 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/61131 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31206 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39112 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84281 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113811 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24925 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99818 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64725 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24281 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/17960 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60690 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/103355 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94309 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18021 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119690 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/109417 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37908 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28166 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93241 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38284 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96092 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93066 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23722 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38114 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15864 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/33882 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37801 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43272 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/133692 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37463 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36101 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40800 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/39169 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hyjorc1 reviewed (approved):
r=me, please remove the changed file list from the commit message if we're not going to comment on those individual files. Adding a statement that this is a refactoring-only patch with no behavior changes would be helpful in the commit message.

Gumichocopengin8 committed:
[JSC] Use `enum class` for `CompilationResult`, `CachedCodeBlockTag`, `CodeSpecializationKind` for Safer C++
https://bugs.webkit.org/show_bug.cgi?id=295459

Reviewed by Yijia Huang and Keith Miller.

This is a refactoring-only patch with no behavior changes.
Use `enum class` for `CompilationResult`, `CachedCodeBlockTag`, `CodeSpecializationKind` for safer C++ [1].

[1]: https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines#use-enum-classes-instead-of-old-style-enumerations

Canonical link: https://commits.webkit.org/297204@main
